reference/opt/shaders-msl/asm/comp/atomic-decrement.asm.comp

```diff
@@ -13,7 +14,8 @@ struct u0_counters
 };
 
 // Returns 2D texture coords corresponding to 1D texel buffer coords
-inline uint2 spvTexelBufferCoord(uint tc)
+static inline __attribute__((always_inline))
```

Comment:
Is it possible for MSL compiler to actually emit function calls here?

Reply:
Which function call do you mean?
The `static inline __attribute__((always_inline))` part is to enforce inlining. Otherwise, linking errors might occur due to duplicate symbols when multiple Metal shaders are linked into one library.

Reply:
I mean, doesn't static inline already make sure functions are inlined? From a linking perspective, inline should fix it. If the toolchain is buggy, I guess we'll have to do this ...

Reply:
I'm afraid the only language "specification" for Metal we have is the [Metal-Shading-Language-Specification.pdf](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf) document, but this doesn't document all compiler features that are inherited from clang (as Metal is a subset of a superset of C++14). As far as I know, even `static inline` is not a guarantee that functions always get inlined. With a quick search I found [this mailing list](https://lore.kernel.org/patchwork/patch/801329/) discussing that compilers don't always honor `inline`. `__attribute__((always_inline))`, however, is a language extension in clang that enforces it, so why not make use of it.

Reply:
Alright, if it's really needed I'm find keeping it for the relevant cases. Not using for templates should remove 90% of them already, so that's nice.